test/e2e/latency.go

```diff
@@ -0,0 +1,277 @@
+		time.Sleep(200 * time.Millisecond)
+	}
+	wg.Wait()
+
+	Logf("Waiting for all Pods begin observed by the watch...")
+	for start := time.Now(); len(watchTimestamps) < nodeCount && time.Since(start) < timeout; time.Sleep(10 * time.Second) {
+	}
+	close(stopCh)
+
+	// Read the schedule timestamp by checking the scheduler event for each pod
+	schedEvents, err := c.Events(ns).List(
+		labels.Everything(),
+		fields.Set{
+			"involvedObject.kind":      "Pod",
+			"involvedObject.namespace": ns,
+			"source":                   "scheduler",
+		}.AsSelector())
+	expectNoError(err)
+	for k := range createTimestamps {
+		for _, event := range schedEvents.Items {
+			if event.InvolvedObject.Name == k {
+				scheduleTimestamps[k] = event.FirstTimestamp
+				break
+			}
+		}
+	}
+
+	var (
+		scheduleLatencies        = make([]podLatencyData, 0)
+		startLatencies           = make([]podLatencyData, 0)
+		watchLatencies           = make([]podLatencyData, 0)
+		scheduleToWatchLatencies = make([]podLatencyData, 0)
+		e2eLatencies             = make([]podLatencyData, 0)
+	)
+
+	for name, podNode := range nodes {
+		createTs, ok := createTimestamps[name]
+		Expect(ok).To(Equal(true))
+		scheduleTs, ok := scheduleTimestamps[name]
+		Expect(ok).To(Equal(true))
+		runTs, ok := startTimestamps[name]
+		Expect(ok).To(Equal(true))
+		watchTs, ok := watchTimestamps[name]
+		Expect(ok).To(Equal(true))
+
+		var (
+			scheduleLatency        = podLatencyData{name, podNode, scheduleTs.Time.Sub(createTs.Time)}
+			startLatency           = podLatencyData{name, podNode, runTs.Time.Sub(scheduleTs.Time)}
+			watchLatency           = podLatencyData{name, podNode, watchTs.Time.Sub(runTs.Time)}
+			scheduleToWatchLatency = podLatencyData{name, podNode, watchTs.Time.Sub(scheduleTs.Time)}
+			e2eLatency             = podLatencyData{name, podNode, watchTs.Time.Sub(createTs.Time)}
+		)
+
+		scheduleLatencies = append(scheduleLatencies, scheduleLatency)
+		startLatencies = append(startLatencies, startLatency)
+		watchLatencies = append(watchLatencies, watchLatency)
+		scheduleToWatchLatencies = append(scheduleToWatchLatencies, scheduleToWatchLatency)
+		e2eLatencies = append(e2eLatencies, e2eLatency)
+	}
+
+	sort.Sort(latencySlice(scheduleLatencies))
+	sort.Sort(latencySlice(startLatencies))
+	sort.Sort(latencySlice(watchLatencies))
+	sort.Sort(latencySlice(scheduleToWatchLatencies))
+	sort.Sort(latencySlice(e2eLatencies))
+
+	printLatencies(scheduleLatencies, "worst schedule latencies")
+	printLatencies(startLatencies, "worst run-after-schedule latencies")
+	printLatencies(watchLatencies, "worst watch latencies")
+	printLatencies(scheduleToWatchLatencies, "worst scheduled-to-end total latencies")
+	printLatencies(e2eLatencies, "worst e2e total latencies")
+
+	// Test whether e2e pod startup time is acceptable.
+	// TODO: Switch it to 5 seconds once we are sure our tests are passing.
+	podStartupThreshold := 8 * time.Second
```

Comment:
The variances for most of these measurements worry me. While I agree having assertions would be a better solution I don't believe there is a stable enough number to choose that would avoid flakes. Since scheduleLatency should be <= 1s I'll add an assertion for it.